packages/api/src/routes/devices/internal-user.ts

```diff
@@ -0,0 +1,62 @@
+import { Response } from "express";
+import Router from "express-promise-router";
+import status from "http-status";
+import { asyncHandler } from "../util";
+import {
+  getAllConnectedUsers,
+  getConnectedUserOrFail,
+} from "../../command/connected-user/get-connected-user";
+import { Constants, providerOAuth2OptionsSchema } from "../../shared/constants";
+import { updateProviderData } from "../../command/connected-user/save-connected-user";
+import { sendProviderDisconnected } from "../../command/webhook/devices";
+
+const router = Router();
+
+/** ---------------------------------------------------------------------------
+ * POST /internal/user/refresh-tokens
+ *
+ * Attempts to refresh all users' tokens for each of their connected providers.
+ * If the refresh doesn't work, this means that the token is invalid and can no
+ * longer be used - in this case, the token will be revoked locally and a
+ * webhook will be sent to the CX to notify them of the disconnect.
+ *
+ *
+ * @return 200 OK
+ */
+router.post(
+  "/refresh-tokens",
+  asyncHandler(async (_, res: Response) => {
+    const connectedUsers = await getAllConnectedUsers();
+    for (const connectedUser of connectedUsers) {
+      const providers = connectedUser.providerMap ? Object.keys(connectedUser.providerMap) : [];
+      const disconnectedProviders: string[] = [];
+      for (const providerStr of providers) {
+        const providerOAuth2Type = providerOAuth2OptionsSchema.safeParse(providerStr);
+        if (providerOAuth2Type.success) {
+          const providerOAuth2 = Constants.PROVIDER_OAUTH2_MAP[providerOAuth2Type.data];
+          try {
+            await providerOAuth2.getAccessToken(connectedUser);
+          } catch (error) {
+            await updateProviderData({
+              id: connectedUser.id,
+              cxId: connectedUser.cxId,
+              provider: providerOAuth2Type.data,
+              providerItem: undefined,
+            });
+            disconnectedProviders.push(providerStr);
+          }
+        }
+      }
+      if (disconnectedProviders) {
+        const newConnectedUser = await getConnectedUserOrFail({
```

Comment:
NABD, since this is admin and likely not going to be used often, but we could store the `connectedUser` w/ the provider name on `disconnectedProviders` and avoid this trip to the DB.

Reply:
I was thinking of this (note [this comment](https://github.com/metriport/metriport/pull/709#discussion_r1272556128)) - opted to just get the actual state from the DB instead of manually deleting from the provider map, and then changing the logic in the wh func